lib/handlers/common/GitHubCredentialsResolver.ts

```diff
@@ -33,58 +32,104 @@ import {
+            if (hasToken(token)) {
+                return { token };
             }
         }
 
-        if (hasToken(this.orgToken)) {
-            return { token: this.orgToken };
-        } else if (hasToken(this.clientToken)) {
-            return { token: this.clientToken };
-        } else if (hasToken(configurationValue("token", "null"))) {
-            return { token: configurationValue<string>("token")};
-        } else if (hasToken(configurationValue("sdm.github.token", "null"))) {
-            return { token: configurationValue<string>("sdm.github.token")};
-        }
+        throw new Error("No GitHub token available! Please add a token to the SDM configuration at 'sdm.github.token' "
+            + "or authenticate the GitHub SCM provider from the web app or CLI.");
+    }
+}
 
-        // Check the graph to see if we have a token on the provider
-        if (!!id && (id.providerType === RepoProviderType.github_com)) {
-            const token = await fetchTokenByProviderType(ProviderType.github_com, context);
-            if (hasToken(token)) {
-                return { token };
-            }
+/**
+ * Obtain a org or user token from the incoming event or command invocation
+ */
+const ObtainTokenFromIncomingMessage: ObtainToken = async ctx => {
+    // Try to obtain the token from the incoming event or command
+    const actx: AutomationContextAware = ctx as any;
+    if (!!actx.trigger && !!actx.trigger.secrets) {
+        let secret = actx.trigger.secrets.find(s => s.uri === Secrets.OrgToken);
+        if (secret && hasToken(secret.value)) {
+            return secret.value;
+        }
+        secret = actx.trigger.secrets.find(s => s.uri.startsWith(Secrets.UserToken));
+        if (secret && hasToken(secret.value)) {
+            return secret.value;
         }
+    }
+    return undefined;
+};
 
-        throw new Error("Neither 'orgToken' nor 'clientToken' has been injected. " +
-            "Please add a repo-scoped GitHub token to your configuration at 'token' or 'sdm.github.token'.");
+/**
+ * Obtain a token from an SCMProvider
+ */
+const ObtainTokenFromProvider: ObtainToken = async (ctx, id) => {
+    // Check the graph to see if we have a token on the provider
+    if (!!id && (id.providerType === RepoProviderType.github_com)) {
+        const token = await fetchTokenByProviderType(ProviderType.github_com, ctx);
+        if (hasToken(token)) {
+            return token;
+        }
     }
+    return undefined;
+};
+
+/**
+ * Obtain a token from the SDM configuration
+ */
+function obtainTokenFromConfiguration(resolver: GitHubCredentialsResolver): ObtainToken {
+    return async () => {
+        if (hasToken(resolver.orgToken)) {
+            return resolver.orgToken;
+        } else if (hasToken(configurationValue("token", "null"))) {
+            return configurationValue<string>("token");
+        } else if (hasToken(configurationValue("sdm.github.token", "null"))) {
```

Comment:
Should the order of token and sdm.github.token be reversed?

Reply:
Not sure. Why do you think it should be reversed? I think there are arguments for both.

Reply:
Typically more specific has higher priority than less specific. But it's not a big deal either way.